### ISSUE TITLE
[go_router] Path and query parameters are not preserved in StatefulShellRoute

### ISSUE THREAD

**Issue by MicTheFox** [flutter/flutter#129878]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

The sample code provided is a variation of the StatefulShellRoute example. In this example the StatefulShellRoute is a SubRoute of the route `/list/:id` which contains the path parameter `:id`. The `:id` parameter is initially set correctly, but is not preserved after calling `goBranch()`. The same behaviour happens to `queryParameters` (not implemented in the provided sample  code).

1. Run the code sample
2. Tap "Section B"

### Expected results

The "id" parameter should stay the same as before.

### Actual results

The "id" parameter actually has the value ":id".

### Code sample

<details><summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:go_router/go_router.dart';

void main() {
  runApp(NestedTabNavigationExampleApp());
}

class NestedTabNavigationExampleApp extends StatelessWidget {
  NestedTabNavigationExampleApp({super.key});

  final GoRouter _router = GoRouter(
    initialLocation: '/list/123/a',
    routes: <RouteBase>[
      GoRoute(
        path: '/list',
        builder: (_, __) => const Placeholder(),
        routes: [
          GoRoute(
            path: ':id',
            builder: (_, __) => const Placeholder(),
            routes: [
              StatefulShellRoute.indexedStack(
                builder: (BuildContext context, GoRouterState state,
                    StatefulNavigationShell navigationShell) {
                  return ScaffoldWithNavBar(navigationShell: navigationShell);
                },
                branches: <StatefulShellBranch>[
                  StatefulShellBranch(
                    routes: <RouteBase>[
                      GoRoute(
                        path: 'a',
                        builder: (BuildContext context, GoRouterState state) =>
                            RootScreen(
                          label: 'A',
                          id: state.pathParameters['id']!,
                        ),
                      ),
                    ],
                  ),
                  StatefulShellBranch(
                    routes: <RouteBase>[
                      GoRoute(
                        path: 'b',
                        builder: (BuildContext context, GoRouterState state) =>
                            RootScreen(
                          label: 'B',
                          id: state.pathParameters['id']!,
                        ),
                      ),
                    ],
                  ),
                ],
              ),
            ],
          ),
        ],
      ),
    ],
  );

  @override
  Widget build(BuildContext context) {
    return MaterialApp.router(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      routerConfig: _router,
    );
  }
}

class ScaffoldWithNavBar extends StatelessWidget {
  const ScaffoldWithNavBar({
    required this.navigationShell,
    Key? key,
  }) : super(key: key ?? const ValueKey<String>('ScaffoldWithNavBar'));

  final StatefulNavigationShell navigationShell;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: navigationShell,
      bottomNavigationBar: BottomNavigationBar(
        items: const <BottomNavigationBarItem>[
          BottomNavigationBarItem(icon: Icon(Icons.home), label: 'Section A'),
          BottomNavigationBarItem(icon: Icon(Icons.work), label: 'Section B'),
        ],
        currentIndex: navigationShell.currentIndex,
        onTap: (int index) => _onTap(context, index),
      ),
    );
  }

  void _onTap(BuildContext context, int index) {
    navigationShell.goBranch(
      index,
      initialLocation: index == navigationShell.currentIndex,
    );
  }
}

class RootScreen extends StatelessWidget {
  const RootScreen({
    required this.label,
    required this.id,
    super.key,
  });

  final String label;
  final String id;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Root of section $label'),
      ),
      body: Center(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            Text('Screen $label',
                style: Theme.of(context).textTheme.titleLarge),
            const Padding(padding: EdgeInsets.all(4)),
            Text('Id: $id'),
          ],
        ),
      ),
    );
  }
}

```

</details>

### Screenshots or Video

<details>
<summary>Screenshots / Video demonstration</summary>

https://github.com/flutter/flutter/assets/138321392/325d48bc-9bdb-456e-bc98-fde88cb9e310

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.10.5, on macOS 13.4.1 22F82 darwin-arm64, locale en-GB)
    • Flutter version 3.10.5 on channel stable at /Users/development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 796c8ef792 (3 weeks ago), 2023-06-13 15:51:02 -0700
    • Engine revision 45f6e00911
    • Dart version 3.0.5
    • DevTools version 2.23.1

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
    • Platform android-33, build-tools 30.0.3
    • Java binary at: /Applications/Android Studio ARM.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.0

[✓] Android Studio (version 2022.1)
    • Android Studio at /Applications/Android Studio ARM.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.15+0-b2043.56-8887301)

[✓] Android Studio (version 2021.1)
    • Android Studio at/Applications/Android Studio ARM.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.11+0-b60-7772763)

[✓] Connected device (3 available)
    • iPhone 14 Pro Max (mobile) • 500A390D-6B90-4F0A-BD61-422F3955A22F • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 13.4.1 22F82 darwin-arm64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 114.0.5735.198

[✓] Network resources
    • All expected network resources are available.

```

</details>

**Comment by darshankawar** [comment]
Thanks for the report @MicTheFox 
Can you take a look at [this](https://github.com/flutter/flutter/issues/123973) issue and see if it resembles your case or not ?

**Comment by Sten435** [comment]
How can this issue be unfixed after 1 month ? Its like a core thing lol

**Comment by d-polikhranidi** [comment]
same behavior with` go_router 12.1.1` and there's still no fix

**Comment by chunhtai** [comment]
this is serious bug, but this should wait for the https://github.com/flutter/packages/pull/5497 to land.

**Comment by firgia** [comment]
I hope the team can fix it ASAP.

I got this issue when I used a 'redirect'.  To fix this issue, I just updated the path from 'redirect' and added custom path parameters and works like a charm.

```
GoRouter(
    routes: [
      GoRoute(
        path: "/email-verified/:email",
        name: RouteName.emailVerified,
        pageBuilder: (context, state) {
          String email = state.pathParameters["email"] ?? "-";
          return CupertinoPage(
            child: EmailVerifiedScreen(
              email: email,
            ),
            fullscreenDialog: true,
          );
        },
      ),
    ],
    redirect: (context, state) {
      String? fullPath = state.fullPath;

      /* UPDATE PATH */
      if (fullPath == "/email-verified/:email") {
        return "/email-verified/${state.pathParameters['email']}";
      }

      return fullPath;
    },
  )
```

**Comment by Uirado** [comment]
Any updates on this issue?

**Comment by bes89** [comment]
@Uirado did you updated go router? For us it was fixed after the update

**Comment by Uirado** [comment]
@bes89 Yes. I tested using version `13.2.0` and this bug is still happening.

**Comment by taylanyildiz** [comment]
Hi, Is there any solution for this

 - Page Detail 
   path: ':id'
   - StatefulShellRoute
      branches: [
       - ShellBranch 
          path: 'properties'
       - ShellBranch
          path: 'notes'

Then when I used the goBranch(index) path parameters returns named like ":id". Same issue [issue#137823](https://github.com/flutter/flutter/issues/137823) I think

**Comment by brunodmn** [comment]
I have the same problem...Thought I might have done something wrong.

I am struggling to create a deeply nested typed navigation with TypedStatefulShellRoute, I have a sidebar menu that navigates to a route that has tabs. Documentation is not quite helping me.

```dart
@TypedStatefulShellRoute<HomeShellRoute>(
  branches: [
    TypedStatefulShellBranch<HomeBranch>(
      routes: <TypedGoRoute<GoRouteData>>[
        TypedGoRoute<HomeRoute>(path: '/home'),
      ],
    ),
   (...),
    TypedStatefulShellBranch<AppsBranch>(
      routes: <TypedGoRoute<GoRouteData>>[
        TypedGoRoute<AppsRoute>(path: '/apps', routes: [
          TypedGoRoute<AppRoute>(path: ':id', routes: [
            TypedStatefulShellRoute<AppShellRoute>(branches: [
              TypedStatefulShellBranch<AppGeneralBranch>(
                  routes: [TypedGoRoute<AppGeneralRoute>(path: 'general')]),
              TypedStatefulShellBranch<AppSupporBranch>(
                  routes: [TypedGoRoute<AppSupportRoute>(path: 'support')]),
            ])
          ]),
        ]),
      ],
    ),
  ],
)
```

![ezgif-3-cd1f8f8708](https://github.com/flutter/flutter/assets/53493535/bce88e5e-c5b0-4b03-b83d-f2b858905d38)

**Comment by P1Rh0** [comment]
Is there any movement on this P1? My BottomDrawer navigation is just broken without it :( is there anything we can do to help?

**Comment by chunhtai** [comment]
I forgot to follow up on this

**Comment by chunhtai** [comment]
The missing path parameters are indeed considered a bug

I am not sure whether the query parameter should carry over by default, what happen if gobranch goes to a branch that has existing state? do we replace it or not?

**Comment by P1Rh0** [comment]
I think I'd probably expect the query parameters not to be passed across multiple routes but would, perhaps, expect to be able to add them optionally via gobranch?
I've not used query parameters myself so happy to bow to anyone who has

**Comment by P1Rh0** [comment]
Just updated and everything looks good to me! Thanks for jumping on this quickly, it's appreciated :)

**Comment by brunodmn** [comment]
@chunhtai  Thank you for your effort and quick reply.

Unfortunately, on my environment, the path parameter is not changing anymore, but nested shell index seems not to change either, which makes my nested navigation stop working.

I did not change any code, just updated dependency go_router to 14.1.2 and dev_dependency go_router_builder to 2.7.0. I ran `flutter pub run build_runner build` after the update and it did not make any changes on my `routes.g.dart` file.


![ScreenRecording2024-05-20at09 53 54-ezgif com-video-to-gif-converter](https://github.com/flutter/flutter/assets/53493535/344a46af-2203-432a-9dff-17d4ba93c7c0)

Its not quite visible on the image, but navigationShell.currentIndex stays on 0, before the updat my onTap event would change it (0 first tab, 1 second tab).
